test/e2e/network/network_policy.go

```diff
@@ -1874,87 +1874,115 @@ var _ = SIGDescribe("NetworkPolicy [Feature:SCTPConnectivity][LinuxOnly][Disrupt
 		framework.DumpDebugInfo(f.ClientSet, f.Namespace.Name)
 
 		pods, policies, logs := collectPodsAndNetworkPolicies(f, podClient)
 		framework.Failf("Pod %s should be able to connect to service %s, but was not able to connect.\nPod logs:\n%s\n\n Current NetworkPolicies:\n\t%v\n\n Pods:\n\t%v\n\n", podClient.Name, service.Name, logs, policies.Items, pods)
-
 	}
 }
 
-func checkNoConnectivity(f *framework.Framework, ns *v1.Namespace, podClient *v1.Pod, service *v1.Service) {
-	framework.Logf("Waiting for %s to complete.", podClient.Name)
-	err := e2epod.WaitForPodSuccessInNamespace(f.ClientSet, podClient.Name, ns.Name)
+func testCannotConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol v1.Protocol, tryCount int) {
+
+	i := 0
+	keepTrying := true
+	var conErr error
+	var podClient *v1.Pod
+
+	// Network policy can take some time to program, especially in scaled/busy environments. Let's allow some retries to
+	// account for that, until we have status field available on network policy object
+
+	for i < tryCount && keepTrying {
+		ginkgo.By(fmt.Sprintf("Creating client pod %s that should not be able to connect to %s.", podName, service.Name))
+		framework.Logf(fmt.Sprintf("Count: %d of tryCount:%d - Creating client pod %s that should not be able to connect to %s.", i+1, tryCount, podName, service.Name))
+		podClient = createNetworkClientPod(f, ns, podName, service, targetPort, protocol)
+		conErr = checkNoConnectivity(f, ns, podClient, service)
+		if conErr != nil {
+			// got the expected result, stop trying
+			keepTrying = false
+		}
+
+		ginkgo.By(fmt.Sprintf("Cleaning up the pod %s", podClient.Name))
+		if err := f.ClientSet.CoreV1().Pods(ns.Name).Delete(context.TODO(), podClient.Name, metav1.DeleteOptions{}); err != nil {
+			framework.Failf("unable to cleanup pod %v: %v", podClient.Name, err)
+		}
+		i++
+	}
 
 	// We expect an error here since it's a cannot connect test.
 	// Dump debug information if the error was nil.
-	if err == nil {
+	if conErr == nil {
 		// Dump debug information for the test namespace.
 		framework.DumpDebugInfo(f.ClientSet, f.Namespace.Name)
 
 		pods, policies, logs := collectPodsAndNetworkPolicies(f, podClient)
 		framework.Failf("Pod %s should not be able to connect to service %s, but was able to connect.\nPod logs:\n%s\n\n Current NetworkPolicies:\n\t%v\n\n Pods:\n\t %v\n\n", podClient.Name, service.Name, logs, policies.Items, pods)
-
 	}
 }
 
+func checkConnectivity(f *framework.Framework, ns *v1.Namespace, podClient *v1.Pod, service *v1.Service) error {
+	framework.Logf("Waiting for %s to complete.", podClient.Name)
+	err := e2epod.WaitForPodNoLongerRunningInNamespace(f.ClientSet, podClient.Name, ns.Name)
+	framework.ExpectNoError(err, "Pod did not finish as expected.")
+
+	framework.Logf("Waiting for %s to complete.", podClient.Name)
+	return e2epod.WaitForPodSuccessInNamespace(f.ClientSet, podClient.Name, ns.Name)
```

Comment:
Oh, people noticed this before but apparently never fixed it; the code here is waiting for the pod to complete twice; the initial `WaitForPodNoLongerRunningInNamespace` is redundant and should just go away. (And then this function is no different from `checkNoConnectivity`.)

Reply:
ok, fixed it, and reduced this down to just checkConnectivity(). Thanks.